src/export/InstanceExporter.ts

```diff
@@ -184,8 +185,53 @@ export class InstanceExporter implements Fishable {
             sourceInfo: rule.sourceInfo
           });
         }
-      } catch (e) {
-        logger.error(e.message, rule.sourceInfo);
+      } catch (originalErr) {
+        // if an Instance has an id that looks like a number, bigint, or boolean,
+        // we may have tried to validate with that value instead of an Instance.
+        // try to fish up an Instance with the rule's raw value.
+        // if we find one, try validating with that instead.
+        if (
+          originalErr instanceof MismatchedTypeError &&
+          (typeof rule.value === 'number' ||
+            typeof rule.value === 'bigint' ||
+            typeof rule.value === 'boolean')
+        ) {
+          const instanceToAssign = this.fishForFHIR(rule.rawValue);
+          if (instanceToAssign == null) {
+            logger.error(originalErr.message, rule.sourceInfo);
+          } else {
+            try {
+              const validatedRule = instanceOfStructureDefinition.validateValueAtPath(
+                rule.path,
+                instanceToAssign,
+                this.fisher,
+                inlineResourceTypes,
+                rule.sourceInfo,
+                manualSliceOrdering
+              );
+              if (validatedRule.pathParts.some(part => part.base.endsWith('[x]'))) {
+                logger.error(
+                  `Unable to assign value at ${rule.path}: choice elements on an instance must use a specific type`,
+                  rule.sourceInfo
+                );
+              } else {
+                ruleMap.set(rule.path, {
+                  pathParts: validatedRule.pathParts,
+                  assignedValue: validatedRule.assignedValue,
+                  sourceInfo: rule.sourceInfo
+                });
+              }
```

Comment:
I was so curious about the looping idea that I tried it and (a) it was kind of messy, and (b) it broke tests. Oh well. But pulling out a function seems OK. I replaced lines 143 - 235 (the body of the `rules.forEach` block) with this and all went well:
```ts
      const inlineResourceTypes: string[] = [];
      // define function that will be re-used in attempting to assign a value or inline instance
      const doRuleValidation = (value: AssignmentValueType) => {
        const validatedRule = instanceOfStructureDefinition.validateValueAtPath(
          rule.path,
          value,
          this.fisher,
          inlineResourceTypes,
          rule.sourceInfo,
          manualSliceOrdering
        );
        // Record each valid rule in a map
        // Choice elements on an instance must use a specific type, so if the path still has an unchosen choice element,
        // the rule can't be used. See http://hl7.org/fhir/R4/formats.html#choice
        if (validatedRule.pathParts.some(part => part.base.endsWith('[x]'))) {
          logger.error(
            `Unable to assign value at ${rule.path}: choice elements on an instance must use a specific type`,
            rule.sourceInfo
          );
        } else {
          ruleMap.set(rule.path, {
            pathParts: validatedRule.pathParts,
            assignedValue: validatedRule.assignedValue,
            sourceInfo: rule.sourceInfo
          });
        }
      };

      try {
        const matchingInlineResourcePaths = inlineResourcePaths.filter(
          i => rule.path.startsWith(`${i.path}.`) && rule.path !== `${i.path}.resourceType`
        );
        // Generate an array of resourceTypes that matches the path, so if path is
        // a.b.c.d.e, and b is a Bundle and D is a Patient,
        // inlineResourceTypes = [undefined, "Bundle", undefined, "Patient", undefined]
        matchingInlineResourcePaths.forEach(match => {
          inlineResourceTypes[splitOnPathPeriods(match.path).length - 1] = match.instanceOf;
        });
        // if a rule has choice elements in the path that are constrained to a single type,
        // change the path to reflect the available type for each choice element.
        // if the path changes, warn the user that it is a better idea to use the more specific element name
        const updatedPath = instanceOfStructureDefinition.updatePathWithChoices(rule.path);
        if (rule.path !== updatedPath) {
          logger.warn(
            `When assigning values on instances, use the choice element's type. Rule path changed from '${rule.path}' to '${updatedPath}'.`,
            rule.sourceInfo
          );
          rule.path = updatedPath;
        }
        doRuleValidation(rule.value);
      } catch (originalErr) {
        // if an Instance has an id that looks like a number, bigint, or boolean,
        // we may have tried to validate with that value instead of an Instance.
        // try to fish up an Instance with the rule's raw value.
        // if we find one, try validating with that instead.
        if (
          originalErr instanceof MismatchedTypeError &&
          ['number', 'bigint', 'boolean'].includes(typeof rule.value)
        ) {
          const instanceToAssign = this.fishForFHIR(rule.rawValue);
          if (instanceToAssign == null) {
            logger.error(originalErr.message, rule.sourceInfo);
          } else {
            try {
              doRuleValidation(instanceToAssign);
            } catch (instanceErr) {
              if (instanceErr instanceof MismatchedTypeError) {
                logger.error(originalErr.message, rule.sourceInfo);
              } else {
                logger.error(instanceErr.message, rule.sourceInfo);
              }
            }
          }
        } else {
          logger.error(originalErr.message, rule.sourceInfo);
        }
      }
```